hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java

```diff
@@ -0,0 +1,266 @@
+      throw new RuntimeException("OM security initialization failed.");
+    case RECOVER:
+      LOG.error("SCM security initialization failed. SCM certificate is " +
+          "missing.");
+      throw new RuntimeException("SCM security initialization failed.");
+    default:
+      LOG.error("SCM security initialization failed. Init response: {}",
+          response);
+      throw new RuntimeException("SCM security initialization failed.");
+    }
+  }
+
+  /**
+   * Get SCM signed certificate and store it using certificate client.
+   */
+  private static void getRootCASignedSCMCert(CertificateClient client,
+      OzoneConfiguration config, String fetchedSCMId,
+      SCMStorageConfig scmStorageConfig, InetSocketAddress scmAddress)
+      throws IOException {
+
+    PKCS10CertificationRequest csr = generateCSR(client, scmStorageConfig,
+        config, scmAddress, fetchedSCMId);
+
+    ScmNodeDetailsProto scmNodeDetailsProto =
+        ScmNodeDetailsProto.newBuilder()
+            .setClusterId(scmStorageConfig.getClusterID())
+            .setHostName(scmAddress.getHostName())
+            .setScmNodeId(fetchedSCMId).build();
+
+    SCMSecurityProtocolClientSideTranslatorPB secureScmClient =
+        HddsServerUtil.getScmSecurityClient(config);
+
+    SCMGetCertResponseProto response = secureScmClient.
+        getSCMCertChain(scmNodeDetailsProto, getEncodedString(csr));
+    String pemEncodedCert = response.getX509Certificate();
+
+    try {
+
+      // Store SCM CA certificate.
+      if (response.hasX509CACertificate()) {
+        String pemEncodedRootCert = response.getX509CACertificate();
+        client.storeCertificate(pemEncodedRootCert, true, true);
+        client.storeCertificate(pemEncodedCert, true);
+
+
+        // Write to the location, so that Default CA server can read the
+        // certificate generated by cert client.
+        CertificateCodec certCodec =
+            new CertificateCodec(new SecurityConfig(config),
+                client.getComponentName());
+
+        X509Certificate certificate =
+            CertificateCodec.getX509Certificate(pemEncodedCert);
+        certCodec.writeCertificate(certCodec.getCertificateHolder(certificate));
+
+        // Persist scm cert serial ID.
+        scmStorageConfig.setScmCertSerialId(certificate.getSerialNumber()
+            .toString());
+      } else {
+        throw new RuntimeException("Unable to retrieve SCM certificate chain");
+      }
+    } catch (IOException | CertificateException e) {
+      LOG.error("Error while storing SCM signed certificate.", e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static void getPrimarySCMSelfSignedCert(CertificateClient client,
+      OzoneConfiguration config, String fetchedSCMId,
+      SCMStorageConfig scmStorageConfig, InetSocketAddress scmAddress)
+      throws IOException {
+
+    PKCS10CertificationRequest csr = generateCSR(client, scmStorageConfig,
```

Comment:
NIT: move generateCSR around line 173 so that if rootCA init failed, no need to generate CSR.

Reply:
Done